man/man8/lconsole.8
.TH LCONSOLE 8
.\" @(#) $Header: /xtel/pp/pp-beta/man/man8/RCS/lconsole.8,v 6.0 1991/12/18 20:44:03 jpo Rel $
.\"
.\" $Log: lconsole.8,v $
.\" Revision 6.0  1991/12/18  20:44:03  jpo
.\" Release 6.0
.\"
.\"
.\"
.SH NAME
lconsole \- line based MTA console
.SH SYNOPSIS
.in +.5i
.ti -.5i
lconsole \%[\-Q\ mta] \%[\-u\ user] \%[\-p\ passwd] \%[\-b] \%[actions...]
.in -.5i
.SH DESCRIPTION
.I Lconsole
is a simple interface to the qmgr providing low functionality but
provides access to all the controlling and monitoring functions.
.PP
The options are:
.TP
\-Q mta
Connect to this mta by default on start up.
.TP
\-u user
Connect as this user to the mta, a password will be prompted for.
.TP
\-p passwd
Supply a password to connect to the mta. (This is not generally
recommended)
.TP
\-b
Run in batch mode (no prompting).
.PP
All other arguments on the command line are taken as an operation to
perform. If no actions are supplied the \fIlconsole\fP takes commands
from the standard input.
.PP
The possible commands are:
.nf
.sp
.ta \w'rereadqueue\0\0'u
?	- provide help.
channel	- display channels or one channel.
clear	- clear the delay on a channel, mta or msg.
close	- close the association with a qmgr.
decrease	- decrease the number of runnable channels.
delay	- add a deley to a channel, mta or msg.
disable	- disable a channel, mta or message.
enable	- enable a channel, mta or msg.
help	- provide help.
increase	- increase the number of runnable channels.
msg	- show details of messages.
mta	- show details of mtas.
open	- open a connection to a qmgr.
quit	- terminate the program.
rereadqueue 	- initiate a queue refresh.
restart	- restart the qmgr.
show	- show certain matching messages.
shutdown	- shutdown the qmgr.
source	- read in a file of commands.
status	- show the status of the qmgr.
.sp
.fi
When displaying channels, messages and mtas, a regular expression is
used to choose the messages. Where a command takes a channel as an
argument, a regular expression resolving to a unique channel may be used.
.SH FILES
None
.SH "SEE ALSO"
MTAconsole(8)
.br
\fIThe PP Manual: Volume 1 \- Installation and Operation\fP
.SH BUGS
Should have more defaulting.


CHANGES









             Changes from PP version 5.2 to 6.0





_1.  _O_v_e_r_v_i_e_w

This  short  document  is  a   description   of   the   more
user/administrator  visible changes that have taken place in
PP between the last interim version (5.2) and the beta  ver-
sion  (6.0). This document is not exhaustive or complete but
is a guide to what to look out for. In general the following
may be assumed.

o+    The Queue files are compatible  between  5.2  and  6.0.
     Note  however that the Queue structure has changed. Old
     channels will have to be recompiled  to  recognise  the
     new format, but new channels will read old queue files.

o+    The qmgr protocol is compatible between 5.2 and 6.0

o+    The syntax of some of  the  table  files  has  changed.
     Where appropriate conversion scripts are provided.

o+    Some procedure calls have been  changed  or  have  dif-
     ferent parameters.

o+    the libdsap.a library is now required for compilation.

In summary, you should be careful when upgrading to the  6.0
release.   It  will  not  work  by  just  replacing  the 5.2
binaries with 6.0. For definite the domain table  will  HAVE
to be changed.

_2.  _D_e_t_a_i_l_e_d _c_h_a_n_g_e_s

This list is in no particular order of preference or  impor-
tance.

Domain Table
     The domain table syntax has been  changed  to  be  more
     explicit  and  has  been  brought into line with the OR
     table syntax. We feel this is clearer and more logical.
     Tools  that  process  the domain table have been taught
     the   new   format   and   there   are    scripts    in
     Tools/tables/compat  to convert between the old and new
     formats.  There is NO backwards compatibility here. Old
     format tables will just plain not work.

Aliases Table



                     December 18, 1991





                           - 2 -


     The aliases table syntax has changed to a  main  choice
     of  synonym or alias with further qualifiers specifying
     the address type of the synonym or alias.  The  changes
     are

          x:alias y         x:alias y
          x:synonym y       x:synonym y
          x:822-synonym y   x:synonym y 822
          x:x400-synonym y  x:synonym y x400
          x:822 y           x:alias y 822
          x:x400 y          x:alias y x400
     where any white space in  y  must  be  quoted  using  a
     backslash  There  is also a "external" qualifier.  This
     disables further lookups of synonyms  and  aliases  and
     can  be  used  for  a  main  site mail hub mapping from
     site_wide_name@localsite                             to
     local_machine_name@localmachine  and  vice  versa.  The
     use of external stops the looping alias  problem  which
     would otherwise occur.

Return of Contents
     Return of contents is now supported.  This  is  a  side
     effect  and  part  of  the  work on routing of delivery
     reports.

Use of pepsy
     Use of the PEPSY ASN.1 compiler rather  than  PEPY/POSY
     is  now  used  in  many  places. This makes for smaller
     binaries and much quicker compilation.

ISODE 7.0
     PP 6.0 is aligned to ISODE 7.0. It will not work  under
     ISODE 6.0.  If you are using the 6.0 release you should
     be using ISODE 7.0. It should however work under  ISODE
     6.8.

Tailor file
     There have been many changes to the  tailor  file.  The
     introduction of several new variables and the splitting
     up of header and body part types. These are

          dap_user     Directory username for DS lookups
          dap_passwd   Directory Passwd for same
          fsync        disable use of fsync(2)
          headertype   list of headers c.f. bodypart
          queuestruct  use hierarchical queue structure
          submit_addr  call submit over network
          diskuse      limit the disk space consumed by PP

     Channel tailoring also has a number of new variables

          drchan       replaces old outchan
          hdrin        headers inbound
          hdrout       headers outbound



                     December 18, 1991





                           - 3 -


          ininfo       inbound channel info
          inadr        inbound channel addr type
          insubadr     inbound channel sub address type
          intable      inbound channel table
          key          give alternative names to channel
          mtatable     list of known mtas for channel binding
          outinfo      outbound channel info
          outsubadr    outbound channel sub address
          outtable     outbound channel table
          solo-proc    channel requires address one at a time
          bad-sender-policyset the policy for unroutable senders
          check        set the checking mode of the channel

Channel pairing
     Channels may now be explicitly  paired  together.  This
     allows one definition to specify both an inbound and an
     outbound channel.  It also allows several logical chan-
     nels  for  one  inbound protocol by selecting a channel
     per MTA.

Timezones
     Many more lettered timezones are recognised. This  list
     is ambiguous at best and numeric forms should really be
     used except at user interfaces.

Delivery Reports
     Delivery reports are now routed back to their  origina-
     tor.  In 5.2 they were returned directly down the chan-
     nel they arrived on. This allows for one way  links  in
     general  makes  things  easier. It also allows protocol
     conversion so that, for instance, a delivery report  on
     an  X.400 message may go back via an RFC-822 if that is
     the only option.  A side effect of this is the  channel
     variable  outchan  has been renamed drchan and has dif-
     ferent semantics attached to it (see the manual).

Header parts
     Separation of headers and bodyparts for  each  channel.
     This  is  mainly a syntactic change but will allow more

Src/lconsole/lconsole.c
/* lconsole.c: driver and main routine for the line console */

# ifndef lint
static char Rcsid[] = "@(#)$Header: /xtel/pp/pp-beta/Src/lconsole/RCS/lconsole.c,v 6.0 1991/12/18 20:27:16 jpo Rel $";
# endif

/*
 * $Header: /xtel/pp/pp-beta/Src/lconsole/RCS/lconsole.c,v 6.0 1991/12/18 20:27:16 jpo Rel $
 *
 * $Log: lconsole.c,v $
 * Revision 6.0  1991/12/18  20:27:16  jpo
 * Release 6.0
 *
 */

#include "lconsole.h"
#include "qmgr-int.h"
#include <varargs.h>

static char	*myname;

int verbose;
int authenticate;
int console_fd = NOTOK;
int accessmode = 0;
int indent = 0;

char *host = NULLCP;
char *realhost = NULLCP;
char *remoteversion = NULLCP;
char *user = NULLCP;
char *passwd = NULLCP;
FILE	*fp_in = stdin;
int	batch = 0;

main (argc, argv)
int	argc;
char	**argv;
{
	extern char	*optarg;
	extern int	optind;
	int	opt;
	char *vec[NVARGS];
	int vecp;
	char *myhost, *mypasswd, *myuser;
	char **op = NULLVP;
	int status;

	myname = argv[0];
	myhost = mypasswd = myuser = NULLCP;
	if (isatty(2))
		verbose = 1;
	while((opt = getopt(argc, argv, "bQ:u:p:")) != EOF)
		switch (opt) {
		    case 'Q':
			myhost = optarg;
			break;

		    case 'u':
			authenticate = TRUE;
			myuser = optarg;
			break;

		    case 'p':
			authenticate = TRUE;
			mypasswd = optarg;
			break;

		    case 'b':
			batch = 1;
			break;

		    default:
			fprintf (stderr, "Usage: %s\n", myname);
			break;
		}
	argc -= optind;
	argv += optind;
	if (argc > 0)
		op = argv;

	if (myhost) {
		vecp = 0;
		vec[vecp++] = "open";
		vec[vecp++] = myhost;
		if (myuser) {
			vec[vecp++] = myuser;
			if (mypasswd)
				vec[vecp++] = mypasswd;
		}
		vec[vecp] = NULLCP;

		if (console_loop (vec, NOTOK) != OK && op)
			exit (1);
	}

	if(op) {
		for (vecp = 0; *op; op ++)
			vec[vecp++] = *op;
		vec[vecp] = NULLCP;

		if (console_loop (vec, NOTOK) != OK)
			status = 1;
	}
	else
		status = interactive ();

	if (console_fd != NOTOK) {
		vecp = 0;
		vec[vecp++] = "close";
		vec[vecp++] = NULLCP;

		(void) console_loop (vec, NOTOK);
	}

	exit(status); /* NOTREACHED */
}

int interactive ()
{
	char *vec[NVARGS];
	int status;
	char buffer[BUFSIZ];
	int i;

	for (;;) {
		if (getline ("%s> ", buffer) == NOTOK)
			break;
		if (buffer[0] == NULL || buffer[0] == '#')
			continue;

		for (i = 0; i < NVARGS; i++)
			vec[i] = NULLCP;
		if (sstr2arg (buffer, NVARGS, vec, " \t") < 1 ||
		    *vec[0] == NULL)
			continue;

		switch (console_loop (vec, OK)) {
		    case NOTOK:
			status = 1;
			break;

		    case OK:
			continue;

		    case DONE:
			status = 0;
			break;
		}
		break;
	}
	return status;
}

int console_loop (vec, retval)
char **vec;
int retval;
{
	struct lc_dispatch *ds;

	if ((ds = getds (*vec)) == NULL)
		return retval;


	if (console_fd == NOTOK) {
		if (ds -> flags & LC_OPEN) {
			advise (NULLCP, "not connected to a queue manager");
			return retval;
		}
	}
	else if (ds -> flags & LC_CLOSE) {
		advise (NULLCP, "already connected to a queue manager");
		return retval;
	}
	if ((ds -> flags & LC_AUTH) && accessmode != FULL_ACCESS) {
		advise (NULLCP, "Operation requires full access rights");
		return retval;
	}

	switch ((*ds -> fnx)(vec)) {
	    default:
	    case NOTOK:
		return retval;

	    case OK:
		return OK;

	    case DONE:
		return DONE;
	}
}


int getline (prompt, buffer)
char *prompt, *buffer;
{
	int c;
	char *cp, *ep;

	if (!batch) {
		printf (prompt, console_fd == NOTOK ? myname : realhost);
		(void) fflush (stdout);
	}

	for (ep = (cp = buffer) + BUFSIZ - 1; (c = getc (fp_in)) != '\n';) {
		if (c == EOF) {
			if (!batch)
				putchar ('\n');
			clearerr(fp_in);
			return NOTOK;
		}
		if (cp < ep)
			*cp ++ = c;
		
	}
	*cp = NULL;
	return OK;
}
#ifndef lint
void    _advise ();


void    adios (va_alist)
va_dcl
{
	va_list ap;

	va_start (ap);

	_advise (ap);

	va_end (ap);

	_exit (1);
}
#else
/* VARARGS */

void    adios (what, fmt)
char   *what,
       *fmt;
{
	adios (what, fmt);
}
#endif


#ifndef lint
void    advise (va_alist)
va_dcl
{
	va_list ap;

	va_start (ap);

	_advise (ap);

	va_end (ap);
}


static void  _advise (ap)
va_list ap;
{
	char    buffer[BUFSIZ];

	asprintf (buffer, ap);

	(void) fflush (stdout);

	fprintf (stderr, "%s: ", myname);
	(void) fputs (buffer, stderr);
	(void) fputc ('\n', stderr);

	(void) fflush (stderr);
}
#else
/* VARARGS */

void    advise (what, fmt)
char   *what,
       *fmt;
{
	advise (what, fmt);
}
#endif

char *datasize (n, units)
int n;
char *units;
{
	static char dbuf[512];

	if (n < 1000)
		(void) sprintf (dbuf, "%d%s", n, units);
	else if (n < 100000)
		(void) sprintf (dbuf, "%.2lfK%s", (double)n/1000.0, units);
	else
		(void) sprintf (dbuf, "%.2lfM%s", (double)n/1000000.0, units);
	return dbuf;
}
